crates/wdk-alloc/src/lib.rs

```diff
@@ -45,52 +50,65 @@ mod kernel_mode {
         // https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-exallocatepool2#remarks
-        let chosen_alignment = if layout.size() >= PAGE_SIZE as usize { PAGE_SIZE as usize } else { MIN_POOL_ALIGNMENT };
+        let chosen_alignment = if layout.size() >= PAGE_SIZE as usize {
+            PAGE_SIZE as usize
+        } else {
+            MIN_POOL_ALIGNMENT
+        };
         // Realignment needed only if requested alignment cannot
         // be satisfied by `ExAllocatePool2`'s chosen alignment
         layout.align() > chosen_alignment
     }
 
-    // SAFETY: This is safe because the Wdk allocator:
-    //         1. can never unwind since it can never panic
-    //         2. has implementations of alloc and dealloc that maintain layout
-    //            constraints (including size and alignment)
     unsafe impl GlobalAlloc for WdkAllocator {
         unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
             let ptr =
-                // SAFETY: `ExAllocatePool2` is safe to call from any `IRQL` <= `DISPATCH_LEVEL` since its allocating from `POOL_FLAG_NON_PAGED`
                 if require_realignment(layout) {
                     // If the requested alignment is too large, we'll have to waste some space as large as the alignment.
                     let size = layout.size() + layout.align();
+                    // Note that `layout.align()` guarantees the alignment will always be a power of 2.
+                    // So we can use the masking trick to find the aligned address.
                     let mask = !(layout.align() - 1);
+                    // SAFETY: `ExAllocatePool2` is safe to call from any `IRQL` <= `DISPATCH_LEVEL`
+                    // since its allocating from `POOL_FLAG_NON_PAGED`
                     let p = unsafe {
                         ExAllocatePool2(POOL_FLAG_NON_PAGED, size as SIZE_T, RUST_TAG)
                     };
                     if !p.is_null() {
                         // Align the pointer up to the first address that meets alignment requirement.
                         let q = ((p as usize & mask) + layout.align()) as *mut PVOID;
-                        // Store the original pointer right before the pointer to return,
-                        // so that ExFreePoolWithTag can receive the correct pointer.
-                        // This is also how `_aligned_malloc` is implemented in msvcrt.dll
+                        // SAFETY: There are sufficient space to store a pointer. See the following graph:
+                        //   | ......... | p | ..................... | ...... |
+                        //                   ^                       ^
+                        //             aligned start            aligned end
+                        // "p" is where we store the original address returned by `ExAllocatePool2`.
+                        // We will also pass "p" to ExFreePoolWithTag on `dealloc` method.
```

Comment:
```suggestion
                        // We will also pass "p" to ExFreePoolWithTag on `dealloc` method.
                        // This is also how `_aligned_free` is implemented in msvcrt.dll
```
I feel we should not lose the comment about `_aligned_free` as it gives useful context.